c10/core/TensorImpl.cpp

```diff
@@ -196,7 +196,11 @@ std::unique_ptr<NamedTensorMetaInterface> NamedTensorMetaInterface::clone() cons
 /// `at::NonVariableTypeMode`.
 #ifndef CAFFE2_FB_LIMITED_MOBILE_CAPABILITY
 
+#ifdef USE_STATIC_DISPATCH
+thread_local bool NonVariableTypeMode_enabled = true;
```

Comment:
Why is variable type mode different depending on static dispatch?

Reply:
For mobile, we're never going through Variable code, but we're passing Variable code through TH methods that expect tensor. Specifically `checked_tensor_unwrap` does an `is_variable()` check, and it'll fail because NonVariableTypeMode is always off, even though we are never going through Variable code. Because our constraints are inference-only and never going through VariableType, I thought it would make sense to just set NonVariableTypeMode and keep it.

@yf225 Do you have any thoughts on this?

Reply:
After @yf225's Variable/Tensor unification work, do we still need keep is_variable() check?

Earlier (before Will landed his work) when I tried removing virtual methods on types & variable, I figured it mostly worked fine without is_variable() check, only need keep a few variable overridden methods virtual:

https://github.com/ljk53/pytorch/commit/2f795a50715f014737d0397f76ed11605939124a
https://github.com/ljk53/pytorch/commit/cc8a1787f79e77cfcff759f92f226afd09feb09a

It's not needed by mobile inference for now but we are still discussing whether we need variable/autodiff for federate on mobile in the future.

Reply:
It's a bit scary - I don't say we will add autograd on mobile build, but coupling with static dispatch like that is suspicious. Can't we just have a lightweight guard (at::AutoNonVariableTypeMode) in every method call?

Probably even better would be to just make sure we don't produce Variables at all - e.g. stub out factory functions if autograd is not compiled in (do we have a dedicated flag for turning off autograd?)